### ISSUE TITLE
BlkExc - Improved peer selection

### ISSUE THREAD

**Issue by benbierens** [codex-storage/nim-codex#1018]
I would like to discuss and design our way to better peer selection in the blockexchange engine. I think we can apply a similar pattern as for IndexingStrategy. We can make an abstraction that hides the peer-selection-strategy from the engine code, so that we can later slot in implementations that optimize for speed vs peer-reliability vs price.

What are the inputs required?
-  Block being downloaded
-  Currently connected peers
Per connected peer we might want to know:
 - Measure of reliability (did they respond to previous messages?)
 - Measure of speed (duration of previous block downloads?)
 - Price for block (can be omitted for now)

The biggest puzzles are acceptable ways of detecting reliability, speed, and deciding how to remember faulty/unreliable peers. At the moment, we just disconnect from them. But if discovery returns them again, we will reconnect.

Please put your insights here.

**Comment by Bulat-Ziganshin** [comment]
A stranger's note: for me, it seems that different users may have very different opinions about the best strategy, and moreover - their personal choice of strategy may be super-important for them (it may directly affect their earnings).

If my assumptions are true, it may be a good place to allow **external** customization, such as a loadable Python script. Not necessarily to implement it ASAP, but you may choose to implement a very simple strategy now, assuming that later you will support external customization and users will soon implement all sorts of strategies and make a "marketplace" of them.

A middle point is to drop an idea of external scripts, but expect that users will drop Nim code with a strategy they've chosen into program sources and compile their own executable. In this case, you just need an extensible API, providing forward and backward compatibility between a strategy code and the main program

**Comment by benbierens** [comment]
I completely agree. This is one very important reason why we should have an abstraction between the engine and the strategy. I suppose initially we will implement in-code our own "PriceOptimizeStrategy" and "SpeedOptimizeStrategy" or similar. But the abstraction should be such that it could support "ExternalPythonScriptStrategy".

A very important question follows: What are the inputs and outputs of such an abstraction? What information can it see, and what does the output look like?
